pkg/retry/retry.go

```diff
@@ -0,0 +1,80 @@
+package retry
+
+import (
+	"context"
+	"time"
+
+	"github.com/sethvargo/go-retry"
+)
+
+const (
+	defaultInterval    = 1 * time.Second
+	defaultMaxRetries  = 10
+	defaultMaxDuration = 60 * time.Second
+)
+
+// RetryConfig is the configuration for a retry operation.
+type RetryConfig struct {
+	// BackoffStrategy is the backoff strategy to use.
+	BackoffStrategy retry.Backoff
+}
+
+// Retryer is a utility for retrying functions.
+type Retryer struct {
+	config *RetryConfig
+}
+
+// NewNoOpRetryer creates a new Retryer that does not retry.
+func NewNoOpRetryer() *Retryer {
+	b := retry.NewConstant(1 * time.Second)
+	b = retry.WithMaxRetries(0, b)
+
+	noOpRetryer := NewRetryer(&RetryConfig{
+		BackoffStrategy: b,
+	})
+
+	return noOpRetryer
+}
+
+func DefaultBackoffStrategy() retry.Backoff {
+	b := retry.NewExponential(1 * time.Second)
+	b = retry.WithMaxDuration(defaultMaxDuration, b)
+	b = retry.WithMaxRetries(defaultMaxRetries, b)
+
+	return b
+}
+
+func NewDefaultRetryer() *Retryer {
+	defaultRetryer := NewRetryer(&RetryConfig{
+		BackoffStrategy: DefaultBackoffStrategy(),
+	})
+
+	return defaultRetryer
+}
+
+// NewRetryer creates a new Retryer with the given configuration.
+func NewRetryer(config *RetryConfig) *Retryer {
+	retryConfig := &RetryConfig{}
+
+	if config != nil {
+		if config.BackoffStrategy != nil {
+			retryConfig.BackoffStrategy = config.BackoffStrategy
+		}
+	} else {
+		retryConfig.BackoffStrategy = retry.NewExponential(defaultInterval)
+	}
+
+	return &Retryer{
+		config: retryConfig,
```

Comment:
Can **retryConfig** ever be just empty? Like **config** is not nil but **config.BackOffStrategy** is?

Reply:
Is that an okay case?

Reply:
good find - I changed the code to not allow this case. now the retryer will use the default configuration in this case. I also added a test